src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/FlatArrayIntervalTree`1.cs

```diff
@@ -0,0 +1,411 @@
+    /// index is beyond the length of this array, that is equivalent to that node not having such a child.
+    /// </summary>
+    /// <remarks>
+    /// The binary tree we represent here is a *complete* binary tree (not to be confused with a *perfect* binary tree).
+    /// A complete binary tree is a binary tree in which every level, except possibly the last, is completely filled,
+    /// and all nodes in the last level are as far left as possible. 
+    /// </remarks>
+    private readonly SegmentedArray<Node> _array;
+
+    private FlatArrayIntervalTree(SegmentedArray<Node> array)
+    {
+        _array = array;
+    }
+
+    /// <summary>
+    /// Provides access to lots of common algorithms on this interval tree.
+    /// </summary>
+    public IntervalTreeAlgorithms<T, FlatArrayIntervalTree<T>> Algorithms => new(this);
+
+    /// <summary>
+    /// Creates a <see cref="FlatArrayIntervalTree{T}"/> from an unsorted list of <paramref name="values"/>.  This will
+    /// incur a delegate allocation to sort the values.  If callers can avoid that allocation by pre-sorting the values,
+    /// they should do so and call <see cref="CreateFromSorted"/> instead.
+    /// </summary>
+    /// <remarks>
+    /// <paramref name="values"/> will be sorted in place.
+    /// </remarks>
+    public static FlatArrayIntervalTree<T> CreateFromUnsorted<TIntrospector>(in TIntrospector introspector, SegmentedList<T> values)
+        where TIntrospector : struct, IIntervalIntrospector<T>
+    {
+        var localIntrospector = introspector;
+        values.Sort((t1, t2) => localIntrospector.GetSpan(t1).Start - localIntrospector.GetSpan(t2).Start);
+        return CreateFromSorted(introspector, values);
+    }
+
+    /// <summary>
+    /// Creates an interval tree from a sorted list of values.  This is more efficient than creating from an unsorted
+    /// list as building doesn't need to figure out where the nodes need to go n-log(n) and doesn't have to rebalance
+    /// anything (again, another n-log(n) operation).  Rebalancing is particularly expensive as it involves tons of
+    /// pointer chasing operations, which is both slow, and which impacts the GC which has to track all those writes.
+    /// </summary>
+    /// <remarks>
+    /// <list type="bullet">The values must be sorted such that given any two elements 'a' and 'b' in the list, if 'a'
+    /// comes before 'b' in the list, then it's "start position" (as determined by the introspector) must be less than
+    /// or equal to 'b's start position.  This is a requirement for the algorithm to work correctly.
+    /// </list>
+    /// <list type="bullet">The <paramref name="values"/> list will be mutated as part of this operation.
+    /// </list>
+    /// </remarks>
+    public static FlatArrayIntervalTree<T> CreateFromSorted<TIntrospector>(in TIntrospector introspector, SegmentedList<T> values)
+        where TIntrospector : struct, IIntervalIntrospector<T>
```

Comment:
this was the code from the binary tree version, moved over her.e

Note: it was changed extensively and should be reviewed.  Specifically, we must generated a *complete* binary tree so that all nodes can find their children through the `2*i + 1 (or) 2` algorithm.  If there are gaps anywhere (say multiple nodes on the second to last level that have a left child, but no right child) then this will break.